### PR TITLE
planner: fix error caused by different length between inner and outer OrderByItems

### LINKED ISSUE #8198: unexpected plan builder error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
``` sql
select * from ((SELECT 1 a,3 b) UNION (SELECT 2,1) ORDER BY (SELECT 2)) t order by a,b;
```

2. What did you expect to see?
``` sql
mysql> select * from ((SELECT 1 a,3 b) UNION (SELECT 2,1) ORDER BY (SELECT 2)) t order by a,b;
+---+---+
| a | b |
+---+---+
| 1 | 3 |
| 2 | 1 |
+---+---+
```


3. What did you see instead?
``` sql
tidb> select * from ((SELECT 1 a,3 b) UNION (SELECT 2,1) ORDER BY (SELECT 2)) t order by a,b;
ERROR 2013 (HY000): Lost connection to MySQL server during query
```
error msg:
``` go
2018/11/06 15:11:14.994 conn.go:427: [error] lastCmd select * from ((SELECT 1 a,3 b) UNION (SELECT 2,1) ORDER BY (SELECT 2)) t order by a,b, runtime error: index out of range, goroutine 20228 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0xc0123c2820, 0xc01686ddff)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:425 +0x10c
panic(0x1e91940, 0x2f468e0)
        /Users/xuhuaiyu/.gvm/gos/go1.11/src/runtime/panic.go:513 +0x1b9
github.com/pingcap/tidb/planner/core.matchItems(0xc008e014f0, 0xc0093633e8, 0x0, 0x1, 0xc00e73e680)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/exhaust_physical_plans.go:748 +0xf3
github.com/pingcap/tidb/planner/core.(*LogicalSort).exhaustPhysicalPlans(0xc00cc72d00, 0xc008e014f0, 0x1, 0xc00c6e08e0, 0x0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/exhaust_physical_plans.go:916 +0x52
github.com/pingcap/tidb/planner/core.(*baseLogicalPlan).findBestTask(0xc00cc72d00, 0xc008e014f0, 0x0, 0x0, 0x0, 0x0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:115 +0x148
github.com/pingcap/tidb/planner/core.(*baseLogicalPlan).findBestTask(0xc00ad7bb98, 0xc008e010e0, 0x0, 0x0, 0x0, 0x0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:122 +0x460
github.com/pingcap/tidb/planner/core.(*baseLogicalPlan).findBestTask(0xc00cc73100, 0xc008e01040, 0x0, 0x0, 0x0, 0x0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/find_best_task.go:122 +0x460
github.com/pingcap/tidb/planner/core.physicalOptimize(0x22221c0, 0xc00cc73100, 0xc00cc73100, 0x22221c0, 0xc00cc73100, 0x0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/optimizer.go:137 +0x9d
github.com/pingcap/tidb/planner/core.DoOptimize(0x113, 0x22221c0, 0xc00cc73100, 0x22221c0, 0xc00cc73100, 0x1, 0x0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/core/optimizer.go:100 +0x7e
github.com/pingcap/tidb/planner.Optimize(0x2222580, 0xc00f656b60, 0x21f84e0, 0xc0071523c0, 0x220be40, 0xc00f365140, 0x2fbf200, 0x0, 0x0, 0x56)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/planner/optimize.go:67 +0x2b6
github.com/pingcap/tidb/executor.(*Compiler).Compile(0xc01686d980, 0x21fd2e0, 0xc00bf3f6c0, 0x21fdfe0, 0xc0071523c0, 0x0, 0x0, 0x0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/executor/compiler.go:49 +0x1d2
github.com/pingcap/tidb/session.(*session).execute(0xc00f656b60, 0x21fd2e0, 0xc00bf3f6c0, 0xc00e365801, 0x56, 0x30, 0x203002, 0x105b1b2, 0x104a2d8, 0x105d140)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/session/session.go:803 +0x6ba
github.com/pingcap/tidb/session.(*session).Execute(0xc00f656b60, 0x21fd2e0, 0xc00bf3f6c0, 0xc00e365801, 0x56, 0x1, 0x0, 0x0, 0xc01686db58, 0x10d7680)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/session/session.go:762 +0x69
github.com/pingcap/tidb/server.(*TiDBContext).Execute(0xc00f65c9c0, 0x21fd2e0, 0xc00bf3f6c0, 0xc00e365801, 0x56, 0xc01686dbe8, 0x1d08b31, 0xc00019bb90, 0x3ff0000000000000, 0x2f9f1a0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/driver_tidb.go:240 +0x7c
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc0123c2820, 0x21fd2e0, 0xc00bf3f6c0, 0xc00e365801, 0x56, 0x0, 0x0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:874 +0x8e
github.com/pingcap/tidb/server.(*clientConn).dispatch(0xc0123c2820, 0xc00e365801, 0x57, 0x57, 0x0, 0x0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:626 +0x655
github.com/pingcap/tidb/server.(*clientConn).Run(0xc0123c2820)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/conn.go:470 +0x1be
github.com/pingcap/tidb/server.(*Server).onConn(0xc006f51ec0, 0x22080c0, 0xc0154bacf0)
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/server.go:324 +0x224
created by github.com/pingcap/tidb/server.(*Server).Run
        /Users/xuhuaiyu/Development/GOPATH/src/github.com/pingcap/tidb/server/server.go:264 +0x4a9

```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
``` sql
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                     |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v2.1.0-rc.3-147-g379ee5b19
Git Commit Hash: 379ee5b1915a37d14e45230acf4d911c6b80d07d
Git Branch: master
UTC Build Time: 2018-11-06 03:12:11
GoVersion: go version go1.11 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by mtunique**
### What problem does this PR solve? 
Fix #8198 nest different size order by.

### What is changed and how it works?
`matchItems` function cannot cover the case that the length of columns is greater than items.
Check the length before iterating the columns.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: planner/core/physical_plan_test.go [hunk 1]
```diff
@@ -189,6 +189,15 @@ func (s *testPlanSuite) TestDAGPlanBuilderSimpleCase(c *C) {
...
 			best: "IndexJoin{TableReader(Table(t)->TopN([test.t.b],0,10))->TopN([test.t.b],0,10)->TableReader(Table(t))}(test.t.a,t1.a)->Limit",
 		},
+		// Test nest different size order by.
+		{
+			sql:  "select * from ((SELECT 1 a,3 b) UNION (SELECT 2,1) ORDER BY (SELECT 2)) t order by a,b",
+			best: "UnionAll{Dual->Projection->Dual->Projection}->HashAgg->Sort",
+		},
+		{
+			sql:  "select * from ((SELECT 1 a,6 b) UNION (SELECT 2,5) UNION (SELECT 2, 4) ORDER BY 1) t order by 1, 2",
+			best: "UnionAll{Dual->Projection->Dual->Projection->Dual->Projection}->HashAgg->Sort->Sort",
```
Comment (eurekaka) [inline planner/core/physical_plan_test.go]: Is it reasonable we have two `Sort`s here?

File: planner/core/physical_plan_test.go [hunk 2]
```diff
@@ -189,6 +189,15 @@ func (s *testPlanSuite) TestDAGPlanBuilderSimpleCase(c *C) {
 			sql:  "select * from (select * from t use index() order by b) t left join t t1 on t.a=t1.a limit 10",
 			best: "IndexJoin{TableReader(Table(t)->TopN([test.t.b],0,10))->TopN([test.t.b],0,10)->TableReader(Table(t))}(test.t.a,t1.a)->Limit",
 		},
+		// Test nest different size order by.
```
Comment (eurekaka) [inline planner/core/physical_plan_test.go]: How about "Test embedded ORDER BY which imposes on different number of columns than outer query"?

File: planner/core/physical_plan_test.go (hunk 1 above)
Comment (mtunique) [reply planner/core/physical_plan_test.go]: Just for test.

**Comment by XuHuaiyu** [comment]
It will be nice if you can add the case mentioned in the issue as an integration test.

**Comment by zz-jason** [comment]
@mtunique Could you please cherry-pick this PR to the release-2.1 and release-2.0 branch after it merged?

**Comment by mtunique** [comment]
> @mtunique Could you please cherry-pick this PR to the release-2.1 and release-2.0 branch after it merged?

OK

**Comment by lzmhhh123** [comment]
LGTM

**Pull request merged**